src/box/box.cc

```diff
@@ -2567,15 +2585,26 @@ box_select(uint32_t space_id, uint32_t index_id,
 		 */
 		space = iterator_space(it);
 	}
-	iterator_delete(it);
 
 	if (rc != 0) {
-		port_destroy(port);
 		txn_rollback_stmt(txn);
-		return -1;
+		goto fail;
 	}
 	txn_commit_ro_stmt(txn, &svp);
+	/*
+	 * We should not update position if no tuple found - if pos was empty,
+	 * it must retain empty, and if no tuple greater then tuple with
+	 * position pos, we must retain it too.
+	 */
+	if (update_pos && index_iterator_position(index, it, &pos->key,
```

Comment:
The index may be dropped in the meantime. We need to handle this. Please take a look at `iterator_is_valid`. Let's raise an error in this case.

Reply:
done